### PR TITLE
[consensus][smoke-tests] Add consensus fault tolerance test

### PR DESCRIPTION
**Pull request by igor-aptos**
### Description

Adding LocalSwarm tests for consensus fault tolerance, and preparing code to be able to run it on remote Forge as well.

### Test Plan
run tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2827)

### CODE REVIEW

**Comment by igor-aptos** [comment]
This is now ready for review. I've split out the utility, so I can use it in forge tests as well.

**Review by zekun000 — APPROVED** [review]

**Comment by igor-aptos** [comment]
oh we need two reviewers to land now. 

@bchocho , @sitalkedia can either one of you take a look?

**Comment by bchocho** [comment]
@igor-aptos taking a look

File: testsuite/forge/src/test_utils/consensus_utils.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        let rounds = cur
+            .iter()
+            .map(|s| s.round)
+            .max()
+            .unwrap()
+            .saturating_sub(previous.iter().map(|s| s.round).max().unwrap());
+        let transactions = cur.iter().map(|s| s.version).max().unwrap()
+            - previous.iter().map(|s| s.version).max().unwrap();
+
+        println!(
```
Comment (bchocho) [inline testsuite/forge/src/test_utils/consensus_utils.rs:100]: Any reason to use println vs info?

**Review by bchocho — COMMENTED** [review]
What happens when we do see a failure? Is there any way to log what failures were triggered when? (e.g., log a seed, or output of actual random values produced).

File: testsuite/smoke-test/src/consensus/consensus_fault_tolerance.rs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+            assert!(
+                executed_transactions >= 4,
+                "no progress with active consensus, only {} transactions",
+                executed_transactions
+            );
+            assert!(
+                executed_rounds >= 2,
+                "no progress with active consensus, only {} rounds",
+                executed_rounds
+            );
```
Comment (bchocho) [inline testsuite/smoke-test/src/consensus/consensus_fault_tolerance.rs:162]: How do you get the assert values? Is this based on cycle duration and round timeout?

File: testsuite/forge/src/test_utils/consensus_utils.rs (hunk 1 above)
Comment (igor-aptos) [reply testsuite/forge/src/test_utils/consensus_utils.rs:100]: no reason, will update

**Comment by igor-aptos** [comment]
@bchocho - this is not a fully controlled test - nodes are started and execute their threads as OS schedules. There isn't a seed to re-run it, unfortunately. 

This will print more info on what went wrong. If used on forge, we will have grafana dashbaords as well.

File: testsuite/smoke-test/src/consensus/consensus_fault_tolerance.rs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        }),
+        true,
+    )
+    .await;
+}
+
+#[tokio::test]
+async fn test_fault_tolerance_of_network_send() {
+    // Randomly increase network failure rate, until network halts, and check that it comes back afterwards.
```
Comment (bchocho) [inline testsuite/smoke-test/src/consensus/consensus_fault_tolerance.rs:171]: At what point does the network halt?

**Review by bchocho — COMMENTED** [review]
Questions for my own understanding

**Review by bchocho — APPROVED** [review]
Anyway looks good :) I don't know if two reviews should be needed for adding tests :P

**Comment by igor-aptos** [comment]
So - at the high level, it is very hard to test properly in the smoke-test, as it needs to run when machine is overloaded.

So most values here are very conservative - i.e. at least some progress, that is like 10x lower than what my machine can do.

but the point is all these tests can just be done over forge/continuous with a more aggressive setups

**Comment by igor-aptos** [comment]
you can also stress-test something locally for much longer. maybe I should add a few such tests under ignore, so people can run when they want to test something more extensively

**Pull request merged**